### PR TITLE
Limit IPAccount Storage Writes to Registered Modules

### LINKED ISSUE #84: Anyone can write to any IPAccounts using their namespace
Anyone can write to any IPAccounts using their namespace (`msg.sender`), for example,
```solidity
function setBytes(bytes32 key, bytes calldata value) external {
    bytesData[_toBytes32(msg.sender)][key] = value;
}
```

We can either
1. Keep this freedom (will attract spams to IPAccount storage)
2. Restrict to permissioned writers, such as registered modules and registries.

Reading will remain 100% public.

### PR DESCRIPTION
**Pull request by kingster-will**
This pull request introduces changes to the IPAccount storage. Here's a brief overview of the changes:

We have now limited the ability to write into the IPAccount storage. With this update, only register modules and core registries will have the write access.

Closes #84

### CODE REVIEW

File: script/foundry/utils/DeployHelper.sol [hunk 1]
```diff
@@ -157,8 +158,16 @@ contract DeployHelper is Script, BroadcastManager, JsonDeploymentHandler, Storag
...
     }
 
+    function create3Deploy(bytes32 salt, bytes calldata creationCode, uint256 value) external returns (address) {
+        return CREATE3.deploy(salt, creationCode, value);
+    }
+
     function _deployProtocolContracts() private {
         require(address(erc20) != address(0), "Deploy: Asset Not Set");
+        bytes32 ipAccountImplSalt = keccak256(
+            abi.encode(type(IPAccountImpl).creationCode, address(this), block.timestamp)
```
Comment (LeoHChen) [inline script/foundry/utils/DeployHelper.sol:168]: the block.timestamp is still random?  do we want to control the salt during the deployment?

**Review by LeoHChen — APPROVED** [review]
LGTM in general, is there any security risk with create3 to deploy the accountimpl? do we want to use the same accountimpl cross multiple chains?

File: contracts/IPAccountStorage.sol [hunk 2]
```diff
@@ -13,15 +15,28 @@ import { ShortString, ShortStrings } from "@openzeppelin/contracts/utils/ShortSt
...
 
     /// @inheritdoc IIPAccountStorage
     function setBytes(bytes32 key, bytes calldata value) external {
+        if (
+            msg.sender != IP_ASSET_REGISTRY &&
+            msg.sender != LICENSE_REGISTRY &&
+            !IModuleRegistry(MODULE_REGISTRY).isRegistered(msg.sender)
+        ) {
+            revert Errors.IPAccountStorage__NotRegisteredModule(msg.sender);
+        }
```
Comment (jdubpark) [inline contracts/IPAccountStorage.sol:39]: Better to have modifier for this check, and use the modifier in both setBytes and setBytes32.

File: test/foundry/IPAccountStorage.t.sol [hunk 3]
```diff
@@ -2,14 +2,18 @@
...
 
 import { IIPAccount } from "../../contracts/interfaces/IIPAccount.sol";
+import { BaseModule } from "../../contracts/modules/BaseModule.sol";
+import { Errors } from "../../contracts/lib/Errors.sol";
 
 import { MockModule } from "./mocks/module/MockModule.sol";
 import { BaseTest } from "./utils/BaseTest.t.sol";
 
-contract IPAccountStorageTest is BaseTest {
+contract IPAccountStorageTest is BaseTest, BaseModule {
```
Comment (jdubpark) [inline test/foundry/IPAccountStorage.t.sol:11]: Why inherit BaseModule for a test?

**Review by jdubpark — COMMENTED** [review]
nit: would be cleaner to check in CREATE3 code in #104

**Comment by LeoHChen** [comment]
please rebase the code.

File: script/foundry/utils/DeployHelper.sol (hunk 1 above)
Comment (kingster-will) [reply script/foundry/utils/DeployHelper.sol:168]: block.timestamp is for making sure the salt is unique; do not have to be random.

File: contracts/IPAccountStorage.sol (hunk 2 above)
Comment (kingster-will) [reply contracts/IPAccountStorage.sol:33]: added modifier.

File: test/foundry/IPAccountStorage.t.sol (hunk 3 above)
Comment (kingster-will) [reply test/foundry/IPAccountStorage.t.sol:11]: To simulate registered module access the ipaccout.

**Review by jdubpark — APPROVED** [review]

**Pull request merged**